src/ConnectedMode/Migration/MSBuildFileProvider.cs

```diff
@@ -31,18 +38,111 @@ namespace SonarLint.VisualStudio.ConnectedMode.Migration
         {
+        }
+
+        internal /* for testing */ MSBuildFileProvider(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            ILogger logger,
+            IThreadHandling threadHandling,
+            IFileSystem fileSystem)
+        {
+            this.serviceProvider = serviceProvider;
             this.logger = logger;
+            this.threadHandling = threadHandling;
+            this.fileSystem = fileSystem;
+        }
+
+        public async Task<IEnumerable<string>> GetFilesAsync(CancellationToken token)
+        {
+            var solutionDir = await GetSolutionDirectoryAsync();
+            if (solutionDir == null)
+            {
+                LogVerbose("Unable to locate the solution folder.");
+                return Enumerable.Empty<string>();
+            }
+
+            // Make sure we are doing the disk search on a background thread
+            await threadHandling.SwitchToBackgroundThread();
+
+            return FindFiles(solutionDir);
+        }
+
+        private async Task<string> GetSolutionDirectoryAsync()
+        {
+            string solutionDir = null;
+
+            await threadHandling.RunOnUIThread(() => {
+                var solution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+                // If there isn't an open solution the returned hresult will indicate an error
+                // and the returned solution name will be null. We'll just ignore the hresult.
+                solution.GetProperty((int)__VSPROPID.VSPROPID_SolutionDirectory, out var solutionDirAsObject);
+
+                solutionDir = solutionDirAsObject as string;
+
+            });
+            return solutionDir;
+        }
+
+        private IEnumerable<string> FindFiles(string rootFolder)
+        {
+            LogVerbose("Searching for files... Root directory: " + rootFolder);
+
+            var timer = Stopwatch.StartNew();
+
+            var allMatches = new HashSet<string>();
+
+            foreach(var pattern in FileSearchPatterns)
+            {
+                var files = fileSystem.Directory.GetFiles(rootFolder, pattern, SearchOption.AllDirectories);
+                var filesToInclude = files.Where(x => !IsInExcludedDirectory(x)).ToArray();
+
+                AddToMatches(allMatches, filesToInclude);
+                LogVerbose($"  Pattern: {pattern}, Number of matching files: {filesToInclude.Length}");
+            }
+
+            timer.Stop();
+            LogVerbose("Total number of matching files: " + allMatches.Count);
+            LogVerbose("Search time (ms): " + timer.ElapsedMilliseconds);
+
+            return allMatches;
         }
 
-        public Task<IEnumerable<string>> GetFilesAsync(CancellationToken token)
+        private static bool IsInExcludedDirectory(string fullPath) => ExcludedDirectores.Any(x => fullPath.Contains(x));
+
+        private static void AddToMatches(HashSet<string> results, IEnumerable<string> matches)
```

Comment:
`HashSet<>` doesn't have an `AddRange` method (unlike `List<>`) so we need this helper method.